### ISSUE TITLE
How to make sure pods startup order within one node by kubelet

### ISSUE THREAD

**Issue by minghuiyuan** [kubernetes/kubernetes#119727]
## What did you expect to happen?
Define some pods start up dependencies, so in one node, if some pods have related dependencies in it. kubelet can follow the dependencies to start up them one by one in order.

for example: network related pods -> monitoring related pods -> application pods -> ...

just like [PriorityClass](https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/#marking-pod-as-critical) with scheduling, if nodes got restarted, some pods should follow the right order to keep that all the environment are set done, and then the normal pods can be startup smoothly.

## What happened?
When cluster nodes got restart with security/maintenance purpose, for example, linux OS patching. Some control plane pods for example:  [istio-cni did not use the hostNetwork, which make the istio start up very slow](https://github.com/istio/istio/issues/46234). if istio start up later than normal pods. which may make all other pods got evicted or crashing. Whenever the nodes get restart by some reasons, the pods may get eviction if some control plane cannot startup in short time.

/sig node

## What did you expect to happen?
Define some pods start up dependencies, so in one node, if some pods have related dependencies in it. kubelet can follow the dependencies to start up them one by one in order.

for example: network related pods -> monitoring related pods -> application pods -> ...

Kubernetes version
v1.23 and master

Cloud provider
None

OS version
Ubuntu22.04

Install tools
None

Container runtime (CRI) and version (if applicable)
N/A

Related plugins (CNI, CSI, ...) and versions (if applicable)
CNI

**Comment by ffromani** [comment]
/triage accepted

II{R,U}C one of the basic design tenets of the platform was that containerized apps are supposed to tolerate restarts and transient error and eventually self-heal when the platform stabilizes.
While this holds true, I think that this issue is another indication that we should probably start considering a way to treat differently infrastructure pods and application pods. We seen something similar in different forms, like https://github.com /kubernetes/kubernetes/issues/109595 and this feels another facet of the same core problem (= we treat infra and app pods equally, we probably should not)

**Comment by minghuiyuan** [comment]
> /triage accepted
> 
> II{R,U}C one of the basic design tenets of the platform was that containerized apps are supposed to tolerate restarts and transient error and eventually self-heal when the platform stabilizes. While this holds true, I think that this issue is another indication that we should probably start considering a way to treat differently infrastructure pods and application pods. We seen something similar in different forms, like https://github.com /[/issues/109595](https://github.com/kubernetes/kubernetes/issues/109595) and this feels another facet of the same core problem (= we treat infra and app pods equally, we probably should not)

yes, for some critical pods, they should be put into the first priority to startup, besides scheduling. Especially for production environment, not only the network related pods, but also for some security guarding purpose pods, all of these sys-infra pods need to get ready first, and then the network environment/policy rules will get ready then, so the application pods will follow the right rule/network-env to work. The feature can solve so many production maintenance related issues that when the cluster/nodes get restart by tech-refresh, ospatching, switch maintenance  and so on. all the pods should not be rescheduled and they just need to restart once on the same node and start up in order.

**Comment by sftim** [comment]
Hmm. We might well want to start pods in order of `priority`, but I disagree that we should track ordering between Pods.

Instead:
- authors should write their app so that it can become healthy, even if Services it depends on aren't available immediately at startup
- DaemonSet Pods should have the right tolerations so that they get started / scheduled when it makes sense to do so
- DaemonSet Pods should drive things towards some kind of desired outcome (eg: working Pod networking; healthy service mesh; CSI driver available on said node)

It's much better for Pods and their containers to solve this problem (eg: a tweak to the Istio code) than to have core Kubernetes try to solve it.

**Comment by ffromani** [comment]
Hey, just thinking if something like https://kubernetes.io/blog/2022/12/26/pod-scheduling-readiness-alpha/ (in principle) for node readines could be a middle ground, making it easier for the correct ordering to happen eventually without changing too much of the kubelet and of the applications.

**Comment by minghuiyuan** [comment]
hi @ffromani , if we follow the principle of https://kubernetes.io/blog/2022/12/26/pod-scheduling-readiness-alpha/ , sounds like the kubelet need to record the tag like "scheduling gate", for example "startupGate". So in one node, kubelet will audit all the pods/containers' score/gates, if one container/pod did have any startupGate, the pods' restart(containers' recreation) may be blocked, that may be a way to let the all the pods/containers startup in order.
